src/combinatorics/simple_graph/basic.lean

```diff
@@ -313,6 +317,72 @@ instance decidable_mem_edge_set [decidable_rel G.adj] :
 instance edges_fintype [decidable_eq V] [fintype V] [decidable_rel G.adj] :
   fintype G.edge_set := subtype.fintype _
 
+/-! ## Darts -/
+
+/-- A `dart` is an oriented edge, implemented as an ordered pair of adjacent vertices. -/
+@[ext, derive decidable_eq]
+structure dart extends V × V :=
+(is_adj : G.adj fst snd)
+
+section darts
+variables {G}
+
+instance dart.fintype [fintype V] [decidable_rel G.adj] : fintype G.dart :=
+fintype.of_equiv (Σ v, G.neighbor_set v)
+{ to_fun := λ s, ⟨(s.fst, s.snd), s.snd.property⟩,
+  inv_fun := λ d, ⟨d.fst, d.snd, d.is_adj⟩,
+  left_inv := λ s, by ext; simp,
+  right_inv := λ d, by ext; simp }
+
+/-- The edge associated to the dart. -/
+def dart.edge (d : G.dart) : sym2 V := ⟦d.to_prod⟧
+
+@[simp] lemma dart.edge_mem (d : G.dart) : d.edge ∈ G.edge_set :=
+d.is_adj
+
+/-- The dart with reversed orientation from a given dart. -/
+def dart.rev (d : G.dart) : G.dart :=
+⟨d.to_prod.swap, G.symm d.is_adj⟩
+
+@[simp] lemma dart.rev_edge : Π (d : G.dart), d.rev.edge = d.edge :=
+by { rintro ⟨⟨v,w⟩, _⟩, exact sym2.eq_swap }
+
+@[simp] lemma dart.rev_rev (d : G.dart) : d.rev.rev = d :=
+by ext; refl
+
+@[simp] lemma dart.rev_involutive : function.involutive (dart.rev : G.dart → G.dart) :=
+dart.rev_rev
+
+lemma dart.rev_ne : Π (d : G.dart), d.rev ≠ d :=
+begin
+  rintro ⟨⟨v, w⟩, h⟩,
+  simp only [dart.rev, not_and, ne.def],
+  rintro ⟨rfl, rfl⟩,
+  exact false.elim (G.loopless _ h),
+end
+
+lemma dart_edge_eq_iff : Π (d₁ d₂ : G.dart),
+  d₁.edge = d₂.edge ↔ d₁ = d₂ ∨ d₁ = d₂.rev :=
+begin
+  rintros ⟨⟨s₁, t₁⟩, h₁⟩ ⟨⟨s₂, t₂⟩, h₂⟩,
+  simp only [dart.edge, dart.rev_edge, dart.rev, sym2.eq_iff, prod.mk.inj_iff, prod.swap_prod_mk],
+end
+
+variables (G)
+
+/-- For a given vertex `v`, this is the bijective map from the neighbor set at `v`
+to the darts `d` with `d.fst = v`. --/
+def dart_of_neighbor_set (v : V) (w : G.neighbor_set v) : G.dart :=
```

Comment:
Note for later, this should probably be called `neighbor_set.dart` (and could take `v` implicit) and be`@[simps]`.

Reply:
I added `@[simps]`, but I'm not sure about `neighbor_set.dart`. If this is for dot notation, a problem is that `w.foo` looks for `coe_sort.foo` unfortunately.